From: Hilko Bengen <bengen@debian.org>
Date: Tue, 9 Feb 2016 23:21:11 +0100
Subject: appliance: Set tar, gzip options for reproducibility

- gzip -n
- explicit file modes inside tarballs
- Set TAR_OPTIONS to supermin prepare invocation in make.sh
---
 appliance/Makefile.am | 8 ++++----
 appliance/make.sh.in  | 3 +++
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/appliance/Makefile.am b/appliance/Makefile.am
index f6dec95..d1ad838 100644
--- a/appliance/Makefile.am
+++ b/appliance/Makefile.am
@@ -55,7 +55,7 @@ superminfs_DATA = \
 supermin.d/base.tar.gz supermin.d/packages: stamp-supermin
 stamp-supermin: make.sh packagelist supermin.d/excludefiles
 	rm -f $@ supermin.d/base.tar.gz supermin.d/packages
-	./make.sh
+	GZIP=-n TAR_OPTIONS="$(TAR_OPTIONS)" ./make.sh
 	$(MAKE) \
 	  supermin.d/daemon.tar.gz \
 	  supermin.d/excludefiles \
@@ -124,7 +124,7 @@ supermin.d/daemon.tar.gz: ../daemon/guestfsd
 	rm -rf tmp-d
 	mkdir -p tmp-d$(DAEMON_SUPERMIN_DIR) tmp-d/etc
 	ln ../daemon/guestfsd tmp-d$(DAEMON_SUPERMIN_DIR)/guestfsd
-	( cd tmp-d && tar $(TAR_OPTIONS) -cf - * ) > $@-t
+	( cd tmp-d && GZIP=-n tar $(TAR_OPTIONS) -cf - * ) > $@-t
 	rm -r tmp-d
 	mv $@-t $@
 
@@ -145,7 +145,7 @@ supermin.d/hostfiles: hostfiles.in Makefile
 
 supermin.d/init.tar.gz: init
 	rm -f $@ $@-t
-	( cd $(srcdir) && tar $(TAR_OPTIONS) -cf - init ) > $@-t
+	( cd $(srcdir) && GZIP=-n tar --mode=0755 $(TAR_OPTIONS) -cf - init ) > $@-t
 	mv $@-t $@
 
 # We should put this file in /lib/udev/rules.d, but put it in /etc so
@@ -155,7 +155,7 @@ supermin.d/udev-rules.tar.gz: 99-guestfs-serial.rules
 	rm -rf tmp-u
 	mkdir -p tmp-u/etc/udev/rules.d
 	for f in $^; do ln $$f tmp-u/etc/udev/rules.d/$$(basename $$f); done
-	( cd tmp-u && tar $(TAR_OPTIONS) -cf - etc ) > $@-t
+	( cd tmp-u && GZIP=-n tar --mode=0644 $(TAR_OPTIONS) -cf - etc ) > $@-t
 	rm -r tmp-u
 	mv $@-t $@
 
diff --git a/appliance/make.sh.in b/appliance/make.sh.in
index ce2231e..dba47db 100755
--- a/appliance/make.sh.in
+++ b/appliance/make.sh.in
@@ -20,6 +20,9 @@ unset CDPATH
 
 set -e
 
+# ensure reproducability
+umask 002
+
 # Run supermin.
 
 if [ "x@SUPERMIN_PACKAGER_CONFIG@" != "xno" ]; then
